File: cses/05_range_queries/1739_forest-queries-ii/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    q = int(data[1])
    W = n + 1
    size = W * W

    bit = [0] * size          # Fenwick 2 chiều, chỉ số phẳng i*W + j (1-indexed)
    cur = bytearray(size)     # trạng thái 0/1 hiện tại của từng ô, để biết chiều toggle
    trans = bytes.maketrans(b'.*', bytes((0, 1)))

    # Nạp lưới ban đầu; gán slice chạy ở tốc độ C.
    for i in range(1, n + 1):
        base = i * W
        seg = data[1 + i].translate(trans)   # data[2] là hàng 1 -> data[1 + i] là hàng i
        bit[base + 1: base + 1 + n] = seg
        cur[base + 1: base + 1 + n] = seg

    # Xây Fenwick "tại chỗ" trong O(n^2): propagate theo hàng, rồi theo cột.
    for i in range(1, n + 1):
        base = i * W
        for j in range(1, n + 1):
            k = j + (j & -j)
            if k <= n:
                bit[base + k] += bit[base + j]
    for j in range(1, n + 1):
        for i in range(1, n + 1):
            k = i + (i & -i)
            if k <= n:
                bit[k * W + j] += bit[i * W + j]

    # Gán biến local để tăng tốc truy cập trong vòng lặp xử lý truy vấn.
    b = bit
    data_local = data
    L = len(data_local)
    idx = 2 + n
    out = []
    ap = out.append

    while idx < L:
        t = data_local[idx]
        if t == b'1':
            # Toggle ô (y, x): xác định delta rồi point-update trên 2D BIT, O(log^2 n).
            y = int(data_local[idx + 1])
            x = int(data_local[idx + 2])
            idx += 3
            p = y * W + x
            if cur[p]:
                cur[p] = 0
                delta = -1
            else:
                cur[p] = 1
                delta = 1
            ii = y
            while ii <= n:
                base = ii * W
                jj = x
                while jj <= n:
                    b[base + jj] += delta
                    jj += jj & -jj
                ii += ii & -ii
        else:
            # Đếm cây trong hình chữ nhật qua bao-trừ 4 góc: Q(y2,x2)-Q(y1-1,x2)-Q(y2,x1-1)+Q(y1-1,x1-1).
            y1 = int(data_local[idx + 1])
            x1 = int(data_local[idx + 2])
            y2 = int(data_local[idx + 3])
            x2 = int(data_local[idx + 4])
            idx += 5
            # Phân tích bit của cột một lần, dùng lại cho mọi hàng trong phân tích của y.
            c2 = []
            jj = x2
            while jj > 0:
                c2.append(jj)
                jj -= jj & -jj
            c1 = []
            jj = x1 - 1
            while jj > 0:
                c1.append(jj)
                jj -= jj & -jj

            res = 0
            # Cộng phần tiền tố tới hàng y2.
            ii = y2
            while ii > 0:
                base = ii * W
                s = 0
                for j in c2:
                    s += b[base + j]
                for j in c1:
                    s -= b[base + j]
                res += s
                ii -= ii & -ii
            # Trừ phần tiền tố tới hàng y1-1.
            ii = y1 - 1
            while ii > 0:
                base = ii * W
                s = 0
                for j in c2:
                    s += b[base + j]
                for j in c1:
                    s -= b[base + j]
                res -= s
                ii -= ii & -ii
            ap(res)

    if out:
        sys.stdout.buffer.write(('\n'.join(map(str, out)) + '\n').encode())


main()

File: cses/05_range_queries/1739_forest-queries-ii/solution.cpp
#include <iostream>
#include <string>
#include <vector>

// Fenwick tree 2 chiều (2D BIT): hỗ trợ point-update và rectangle-sum trên lưới.
class Fenwick2D {
public:
    Fenwick2D(int n, const std::vector<std::string>& grid)
        : n_(n), width_(n + 1), data_(static_cast<std::size_t>(width_ * width_), 0) {
        // Nạp giá trị 0/1 của lưới vào mảng (1-indexed).
        for (int row = 1; row <= n_; ++row) {
            for (int column = 1; column <= n_; ++column) {
                data_[index(row, column)] =
                    grid[static_cast<std::size_t>(row)][static_cast<std::size_t>(column - 1)] == '*';
            }
        }

        // Xây Fenwick "tại chỗ" trong O(n^2): propagate lần lượt theo hàng, rồi theo cột.
        for (int row = 1; row <= n_; ++row) {
            for (int column = 1; column <= n_; ++column) {
                const int parent = column + (column & -column);
                if (parent <= n_) {
                    data_[index(row, parent)] += data_[index(row, column)];
                }
            }
        }
        for (int column = 1; column <= n_; ++column) {
            for (int row = 1; row <= n_; ++row) {
                const int parent = row + (row & -row);
                if (parent <= n_) {
                    data_[index(parent, column)] += data_[index(row, column)];
                }
            }
        }
    }

    // Point-update: cộng value vào ô (row, column) qua hai vòng Fenwick lồng nhau, O(log^2 n).
    void add(int row, int column, int value) {
        for (int i = row; i <= n_; i += i & -i) {
            for (int j = column; j <= n_; j += j & -j) {
                data_[index(i, j)] += value;
            }
        }
    }

    // Tổng hình chữ nhật theo công thức bao-trừ (inclusion-exclusion) 4 góc.
    [[nodiscard]] int rectangle_sum(int row1, int column1, int row2, int column2) const {
        return prefix_sum(row2, column2) - prefix_sum(row1 - 1, column2) -
               prefix_sum(row2, column1 - 1) + prefix_sum(row1 - 1, column1 - 1);
    }

private:
    int n_;
    int width_;
    std::vector<int> data_;

    // Ánh xạ ô (row, column) tới chỉ số phẳng trên mảng.
    [[nodiscard]] std::size_t index(int row, int column) const {
        return static_cast<std::size_t>(row * width_ + column);
    }

    // Tổng tiền tố của hình chữ nhật (1,1)..(row,column), O(log^2 n).
    [[nodiscard]] int prefix_sum(int row, int column) const {
        int result = 0;
        for (int i = row; i > 0; i -= i & -i) {
            for (int j = column; j > 0; j -= j & -j) {
                result += data_[index(i, j)];
            }
        }
        return result;
    }
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n = 0;
    int q = 0;
    std::cin >> n >> q;
    std::vector<std::string> grid(static_cast<std::size_t>(n + 1));
    for (int row = 1; row <= n; ++row) {
        std::cin >> grid[static_cast<std::size_t>(row)];
    }

    Fenwick2D tree(n, grid);
    while (q-- > 0) {
        int type = 0;
        std::cin >> type;
        if (type == 1) {
            // Toggle ô (row, column): nếu đang là cây thì bớt 1, ngược lại thêm 1.
            int row = 0;
            int column = 0;
            std::cin >> row >> column;
            char& cell = grid[static_cast<std::size_t>(row)][static_cast<std::size_t>(column - 1)];
            if (cell == '*') {
                cell = '.';
                tree.add(row, column, -1);
            } else {
                cell = '*';
                tree.add(row, column, 1);
            }
        } else {
            // Đếm số cây trong hình chữ nhật (row1, column1)..(row2, column2).
            int row1 = 0;
            int column1 = 0;
            int row2 = 0;
            int column2 = 0;
            std::cin >> row1 >> column1 >> row2 >> column2;
            std::cout << tree.rectangle_sum(row1, column1, row2, column2) << '\n';
        }
    }
}
